src/util/sss_regexp.c

```diff
@@ -0,0 +1,215 @@
+
+#ifdef HAVE_LIBPCRE2
+/*
+ * sss_regexp with pcre2
+ */
+struct _sss_regexp_t {
+    pcre2_code *re;
+    unsigned char *error_msg;
+    pcre2_match_data *match_data;
+    char *matched_string;
+};
+
+int sss_regexp_pcre2_destroy(sss_regexp_t *self)
+{
+    if (self->re) pcre2_code_free(self->re);
+    if (self->error_msg) free (self->error_msg);
+    if (self->match_data) pcre2_match_data_free(self->match_data);
+    if (self->matched_string) pcre2_substring_free((PCRE2_UCHAR *)self->matched_string);
+    return 0;
+}
+
+void sss_regexp_pcre2_compile(sss_regexp_t *self,
+                              const char *pattern,
+                              int options,
+                              int *errorcodeptr,
+                              const char **errormsgptr,
+                              size_t *erroffset)
+{
+    self->re = pcre2_compile((PCRE2_SPTR)pattern, strlen(pattern), options, errorcodeptr, erroffset, NULL);
+    self->error_msg = malloc(SSS_REGEXP_ERR_MSG_SIZE);
+    if (self->re && self->error_msg) {
+        pcre2_get_error_message(*errorcodeptr, self->error_msg, SSS_REGEXP_ERR_MSG_SIZE);
+    }
+}
+
+int sss_regexp_pcre2_match(sss_regexp_t *self,
+                           const char *subject,
+                           int startoffset,
+                           int options)
+{
+    if (self->match_data) pcre2_match_data_free(self->match_data);
+    self->match_data = pcre2_match_data_create_from_pattern(self->re, NULL);
+    return pcre2_match(self->re, (PCRE2_SPTR)subject, strlen(subject), startoffset, options, self->match_data, NULL);
+}
+
+int sss_regexp_pcre2_get_named_substring(sss_regexp_t *self,
+                                         const char *name,
+                                         const char **value)
+{
+    PCRE2_SIZE length;
+
+    if (self->matched_string) {
+        pcre2_substring_free((PCRE2_UCHAR *)(self->matched_string));
+        self->matched_string = NULL;
+    }
+    int rc = pcre2_substring_get_byname(self->match_data, (PCRE2_SPTR)name, (PCRE2_UCHAR **) &self->matched_string, &length);
+    *value = self->matched_string;
+    return rc;
+}
+
+#else /* HAVE_LIBPCRE2 */
+/*
+ * sss_regexp with pcre
+ */
+struct _sss_regexp_t {
+    pcre *re;
+    int ovector[SSS_REGEXP_OVEC_SIZE];
+    const char *matched_string;
+    const char *subject;
+};
+
+int sss_regexp_pcre1_destroy(sss_regexp_t *self)
+{
+    if (self->re) pcre_free(self->re);
+    if (self->matched_string) pcre_free_substring(self->matched_string);
+    return 0;
+}
+
+void sss_regexp_pcre1_compile(sss_regexp_t *self,
+                              const char *pattern,
+                              int options,
+                              int *errorcodeptr,
+                              const char **errormsgptr,
+                              size_t *erroffset)
+{
+    self->re = pcre_compile2(pattern, options, errorcodeptr, errormsgptr, (int *)erroffset, NULL);
```

Comment:
Here I also think the compilation result could be checked.

Reply:
I don't see any benefit in checking result here. It must be (is) checked in place where we call this function from to log an error and so on.